Check manually-assigned addresses with IPAM to see if they clash

bboreham opened the pull request:
Fixes #687 and #598

One quiet change in this PR:
 * avoid printing error messages from 'claim' twice in the logs

rade commented:
> I could add some more lines to remove this if desired.

yes, please.

bboreham commented:
Done

squaremo commented on ipam/allocator.go:
```diff
@@ -215,9 +215,9 @@ func (alloc *Allocator) Lookup(ident string, r address.Range) (address.Address,
+func (alloc *Allocator) Claim(ident string, addr address.Address, launch bool) error {
 	resultChan := make(chan error)
-	op := &claim{resultChan: resultChan, ident: ident, addr: addr}
+	op := &claim{resultChan: resultChan, ident: ident, addr: addr, launch: launch}
```
I do not understand what `launch` mean in these changes

bboreham replied:
I was trying to convey that we want it to behave differently when called at launch of weave compared to later.

squaremo replied:
I think it would be better to name it for how it behaves differently

squaremo commented:
I think these are two different operations, rather than a variation on one operation (and perhaps that is why I was confused by "launch").

The first is to tell weave about an IP that had been claimed previously. The second is to ask for a particular IP that has not been handed out previously. It's actually the latter that deserves the name "Claim"; the former is more like, well, "Reclaim" or "Recover".

How this is reflected in the HTTP API, I'm not sure -- PUT to a URL seems a good match for both. So a query arg might be appropriate for discriminating.

bboreham commented:
I renamed 'launch' to 'reclaim'; rebased and squashed.
Also changed the behaviour to output a message and carry on when IPAM is waiting for quorum, rather than hang the caller.  This is more in line with what #687 asks for.

squaremo commented on ipam/claim.go:
```diff
@@ -12,6 +12,7 @@ type claim struct {
+	reclaim    bool
 }
 
 func (c *claim) sendResult(result error) {
```
As a drive-by comment: I find it somewhat counter-intuitive that `claim.sendResult(nil)` indicates a successful operation.

squaremo commented on ipam/claim.go:
```diff
@@ -40,15 +44,23 @@ func (c *claim) Try(alloc *Allocator) bool {
-			alloc.infof("Claim %s for %s: %s; will try later.", c.addr, c.ident, err)
-			c.sendResult(nil)
+		if err != nil { // can't speak to owner right now
+			if c.reclaim { // don't hold up launch; figure it out later
```
> // don't hold up launch; figure it out later

This comment doesn't make sense -- what does `reclaim` have to do with launch?

bboreham replied:
this code is only called from `weave launch`

squaremo replied:
that's a coincidence

squaremo replied:
Or think of it this way: if reclaim is a general-purpose operation that happens to be called at launch, this comment doesn't belong here. If reclaim is an operation intended _only_ for use during launch, why is it not better adapted to its purpose, e.g., a bulk operation (in other words, why does it look general purpose ..)

bboreham replied:
that's an accident

squaremo replied:
> that's an accident

which bit?

bboreham replied:
that it looks general-purpose.  Certainly the behaviour of apparently agreeing with a claim when we can't answer for sure yet (introduced in #1104) is purely directed at launch-time.

squaremo replied:
OK. As it is, it's a non-sequitur: "if reclaiming, don't hold up launch". Why should you have to know how it's used in the script to be able to make sense of that?

As much as I dislike the idea of holding things up for a comment, I think this is going to cause befuddlement for the next person to look at the bit of code. What I'd rather see is an explanation of why there is a different behaviour for an apparently very similar operation -- or for it to obviously be a different operation.

bboreham commented:
I decided `noErrorOnUnknown` described the behaviour best.

squaremo commented on ipam/claim.go:
```diff
@@ -40,15 +43,23 @@ func (c *claim) Try(alloc *Allocator) bool {
+			alloc.infof("Claim %s for %s: address allocator still initializing; will try later.", c.addr, c.ident)
+			c.sendResult(nil) // don't make the caller wait
+		} else {
+			c.sendResult(fmt.Errorf("%s is in the range %s, but the allocator is not initialized yet", c.addr, alloc.universe.AsCIDRString()))
```
The suggestion of the previous comment is that this will block the caller. Is that the case?

squaremo replied:
Whatever, I'll fix it up in post

squaremo replied:
> Whatever, I'll fix it up in post

#1219

rade commented:
I struggle to see how this solves #598. Please explain.

bboreham commented:
Line 866 in the `weave` script informs IPAM of a manually-selected address.  If this can be granted, it is, and hence IPAM will not then allocate that address to something else, which I thought was the main issue of #598.

rade commented:
ok

bboreham commented:
I should perhaps have updated the docs, where it says 

> make sure they don’t conflict with any IP ranges in use on the hosts (including those delegated to weave’s automatic IP address allocator)

rade commented:
Surely http://docs.weave.works/weave/master/head/ipam.html#manual should be updated.
